src/hotspot/share/classfile/classLoader.cpp

```diff
@@ -811,6 +813,7 @@ void ClassLoader::add_to_app_classpath_entries(JavaThread* current,
     while (e != nullptr) {
       if (strcmp(e->name(), entry->name()) == 0) {
         // entry already exists
+        delete entry;
```

Comment:
I think the API contract between `update_class_path_entry_list()` and `add_to_app_classpath_entries()` is too complicated:

- `add_to_app_classpath_entries()` will delete the entry if it decides that the entry shouldn't be saved

It's better for the lifetime of the entry to be maintained by `update_class_path_entry_list()`. It should do something like:

```
 new_entry = create_class_path_entry(current, path, &st, is_boot_append, from_class_path_attr);

 if (add_to_app_classpath_entries(current, new_entry, check_for_duplicates) == false) {
  // new_entry is not saved, free it now
  delete new_entry;
 }
```

Reply:
Thanks for the review. I've pushed another commit based on your suggestions.